### ISSUE TITLE
<If> creates <span> if not executed within <For> loop

### ISSUE THREAD

**Issue by texttechne** [AlexGilleran/jsx-control-statements#17]
If the `<If ...>` statement is not executed, then an empty `<span>` is rendered, due to the fact that an empty string is returned: See https://github.com/AlexGilleran/jsx-control-statements/blob/master/jstransform.js#L79

Would be great if you would consider changing this empty string to `null`, so that nothing is rendered.

Problematic cases are those where span tags are not allowed, e.g. within `<tr>` ord `<td>`. React throws warnings in this case...

**Comment by AlexGilleran** [comment]
Huh, ok. I seem to recall that the reason this was done originally was that returning `null` broke parsing, but an empty string did nothing. Annoying if that's changed, although possibly I got it wrong in the beginning.

**Comment by nuclearspike** [comment]
+1 for getting warnings about spans between tr/td Glad to see this is already listed as a bug

**Comment by texttechne** [comment]
The problem is more complicated then stated above. Minimal test case:
```
<div>
  <For ...>
    <If condition={false}>
      Not Rendered
    </If>
  </For>
</div>
```
My suggestion to use `null` instead of an empty string works for `jstransform`: No empty `<span>` element is rendered then. 

However, there are a couple of strange things going on which I don't understand. 
* Why are no spans rendered when `<If>` is not wrapped inside of a for loop, i.e. without the mentioned fix? 
* If using babel and running your test suite, then within `babel.js` the essential function `transformIf` is not executed for even the most simple tests. I can comment out the whole function and only the error tests fail... => So how is the transform for Babel realized???

I would like to submit a PR with test case for this, but I don't get how the Babel plugin works... @AlexGilleran  Could you shed some light on this?

**Comment by AlexGilleran** [comment]
I can definitely answer the 2nd one re: the babel tests - The JSTransform and Babel test suites have to be executed completely separately. If you try to run them both at the same time, JSTransform gets registered as the transformer for JSX files and doesn't get un-registered (I could never find a reliable way to do this) when the Babel tests start, which is why only the Babel-specific tests (the error ones) fail. Sorry, this should probably be in a contributors guide.

As for the first question, I'm not sure, although at least that explains why I thought an empty string wouldn't create a span. I guess it's because React interprets `''` as no content, but interprets `['', '', '']` as [`<span>`, `<span>`, `<span>`]?

**Comment by texttechne** [comment]
Yeah, you're right about the first question: 
* empty string => nothing is rendered
* list of empty strings => empty `span`'s are rendered

The second question, however, remains the crucial and bugging one. Figured already out that JSTransform and Babel test suites have to be executed separately, before I've asked you. Double checked, that JSTransform does not get executed (even deleted the file). This is how I run the babel tests:
```mocha spec/with-babel/*.js```

However, no method from `babel.js` gets executed for the main tests (only for the error ones...). I can even go as far as to reduce `babel.js` to only return an empty object (sic!) and still something transforms those `if` and `for` statements (even if neither `jstransform.js` nor the matching node modules exist within the repo). The only thing I have to do is to reference babel.js (as said: returning only an empty object) as plugin, otherwise if and for statements are unknown... I'm absolutely baffled...

**Comment by texttechne** [comment]
Undocumented caching, yippie! Babel caches all compiled stuff in home directory. Cache busting can be enabled for Babel by specifying `cache: false` via require hook or `.babelrc`. Created some pull requests...

**Comment by AlexGilleran** [comment]
Heh, how very Babel.

PRs look great, thanks. Will have a proper look/merge when I get home tonight :+1:

**Comment by AlexGilleran** [comment]
Fixed by @texttechne, thanks :smiley:

**Comment by AlexGilleran** [comment]
Published in 2.0.2
